lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -79,17 +79,25 @@ sub ajax {
             my @data;
             my %modules;
 
-            my %needle_dir_map = map { $_->path => $_ } $needles->all;
-
+            my $schema = $self->schema;
             while (my $n = $needles->next) {
-                my $hash = {
+                my $filename = $n->filename;
+                my $hash     = {
                     id        => $n->id,
                     directory => $n->directory->name,
-                    filename  => $n->filename,
+                    filename  => $filename,
                 };
+                my $needle_dir           = $n->directory->path;
+                my $needle_dir_real_path = realpath($needle_dir);
+                if ($needle_dir_real_path && $needle_dir_real_path ne $needle_dir) {
+                    my $real_needle_info;
+                    my $real_needle_dir = $self->schema->resultset('NeedleDirs')->find({path => $needle_dir_real_path});
+                    $real_needle_info
+                      = $schema->resultset('Needles')->find({dir_id => $real_needle_dir->id, filename => $filename})
+                      if $real_needle_dir;
```

Comment:
so for every single needle you want to access the DB two times additionally? This looks like it can have severe performance implications

Reply:
If the needle is a symlink, I need to access the DB two times to get the real needle directory id and real needle information. I have not tested the performance. I guess the web page only load 10 (max is 100) needles at one time, maybe the performance implication will not too bad. I will test it.

Reply:
I modify the code to record the directory id. This will reduce the times of accessing the DB if the needles on one page belong to the same directory.